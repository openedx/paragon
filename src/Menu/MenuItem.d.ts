import React from 'react';

interface MenuItemProps extends React.HTMLAttributes<HTMLElement> {
  defaultSelected?: boolean;
  children?: React.ReactNode;
  as?: keyof JSX.IntrinsicElements;
  iconBefore?: React.ReactNode;
  iconAfter?: React.ReactNode;
  disabled?: boolean;
  href?: string;
  variant?: string;
  size?: string;
}

declare const MenuItem: React.FC<MenuItemProps>;

export default MenuItem;
