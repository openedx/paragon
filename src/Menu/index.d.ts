import * as React from 'react';

export interface IMenu {
  className?: string;
  arrowKeyNavigationSelector?: string;
  as?: keyof JSX.IntrinsicElements;
  children?: React.ReactNode;
}

declare const Menu: React.FC<IMenu>;

export default Menu;
