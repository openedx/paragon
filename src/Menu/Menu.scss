@import "variables";

.pgn__menu {
  border-radius: $menu-border-radius;
  box-shadow: $menu-box-shadow;
  background-color: $menu-bg;
  overflow: auto;
  max-height: $menu-max-height;

  .pgn__menu-item {
    display: flex;
    align-items: center;
    justify-content: flex-start;
    width: $menu-item-width;
    color: $menu-item-color;
    text-align: center;
    vertical-align: middle;
    user-select: none;
    background-color: $menu-item-bg;
    border: $menu-item-border;
    padding: $menu-item-padding-y $menu-item-padding-x;
    height: $menu-item-height;

    @media (max-width: $max-width-xs) {
      width: $menu-item-width-xs;
    }

    .pgn__menu-item-text {
      display: block;
      text-overflow: ellipsis;
      white-space: nowrap;
      overflow: hidden;
    }

    .pgn__menu-item-content-spacer {
      flex-grow: 1;
    }

    @include hover {
      text-decoration: none;
      color: $menu-item-hover-color;
      border-color: $menu-item-hover-border-color;
      background: $menu-item-hover-bg;
    }

    // Disabled comes first so active can properly restyle
    &.disabled,
    &:disabled {
      opacity: $btn-disabled-opacity;
    }

    &.focus,
    &:focus {
      font-weight: $font-weight-bolder;
      background-color: $menu-item-focus-bg;
    }

    .btn-icon-before {
      margin-inline-end: .5rem;
      margin-left: calc(#{$menu-item-icon-margin-left} * -1);

      [dir="rtl"] & {
        margin-right: calc(#{$menu-item-icon-margin-right} * -1);
        margin-left: .5rem;
      }
    }

    .btn-icon-after {
      margin-inline-start: .5rem;
      margin-right: calc(#{$menu-item-icon-margin-right} * -1);

      [dir="rtl"] & {
        margin-right: .5rem;
        margin-left: calc(#{$menu-item-icon-margin-left} * -1);
      }
    }
  }
}

.pgn__menu-select {
  .btn.btn-link {
    color: $menu-select-btn-link-color;

    &:hover,
    &:focus {
      text-decoration-line: $menu-select-btn-link-text-decoration-line;
      text-decoration-thickness: $menu-select-btn-link-text-decoration-thickness;
    }
  }

  .pgn__menu-select-popup {
    position: relative;
  }
}
