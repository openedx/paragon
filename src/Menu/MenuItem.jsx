import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import Icon from '../Icon';

function MenuItem({
  as,
  children,
  defaultSelected,
  iconAfter,
  iconBefore,
  ...props
}) {
  const className = classNames(props.className, 'pgn__menu-item');

  return React.createElement(
    as,
    {
      ...props,
      className,
    },
    (
      <>
        {iconBefore && <Icon className="btn-icon-before" src={iconBefore} />}
        <span className="pgn__menu-item-text">{children}</span>
        <span className="pgn__menu-item-content-spacer" />
        {iconAfter && <Icon className="btn-icon-after" src={iconAfter} />}
      </>
    ),
  );
}

MenuItem.propTypes = {
  /** Specifies that this ``MenuItem`` is selected inside the ``SelectMenu`` */
  defaultSelected: PropTypes.bool,
  /** Specifies class name to append to the base element */
  className: PropTypes.string,
  /** Specifies the content of the ``MenuItem`` */
  children: PropTypes.node,
  /** Specifies the base element */
  as: PropTypes.elementType,
  /** Specifies the jsx before the content of the ``MenuItem`` */
  iconBefore: PropTypes.oneOfType([PropTypes.element, PropTypes.func]),
  /** Specifies the jsx after the content of the ``MenuItem`` */
  iconAfter: PropTypes.oneOfType([PropTypes.element, PropTypes.func]),
};

MenuItem.defaultProps = {
  defaultSelected: false,
  as: 'button',
  className: undefined,
  children: null,
  iconBefore: undefined,
  iconAfter: undefined,
};

export default MenuItem;
