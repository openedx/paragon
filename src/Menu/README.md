---
title: 'Menu'
type: 'component'
components:
- Menu
- MenuItem
categories:
- Navigation
status: 'New'
designStatus: 'Done'
devStatus: 'Done'
notes: ''
---

## Menu

An arrow-key navigable ``Menu`` which consists of ``MenuItems``. A ``Menu`` can be employed to produce its common variants, including dropdown menus, select menus, and others. ``Menus`` are keyboard navigable with both tab and arrow keys.

## MenuItem

A ``MenuItem`` is its own distinct component that is used by any kind of menu overlays i.e. dropdown menu and navigation menu.

```jsx live
() => {
  return (
    <Menu>
      <MenuItem> A Menu Item</MenuItem>
      <MenuItem iconBefore={Check}>A Menu Item With an Icon Before</MenuItem>
      <MenuItem iconAfter={Check}>A Menu Item With an Icon After </MenuItem>
      <MenuItem disabled>A Disabled Menu Item</MenuItem>
      <MenuItem as={Hyperlink} destination="https://en.wikipedia.org/wiki/Hyperlink">A Link Menu Item</MenuItem>
      <MenuItem as={Button} variant="tertiary" size="inline">A Button Menu Item</MenuItem>
      <MenuItem as={Form.Checkbox}>A Checkbox Menu Item</MenuItem>
    </Menu>
  );
}
```

## With Form

A ``Menu`` can include things like forms.

```jsx live
() => {
  return (
    <Form.Group>
      <Form.CheckboxSet
          name="color-two"
          onChange={(e) => console.log(e.target.value)}
          defaultValue={['green']}
        >
        <Form.Label>Which Color?</Form.Label>
        <Menu>
          <MenuItem as={Form.Checkbox} value="red">Red</MenuItem>
          <MenuItem as={Form.Checkbox} value="blue">Blue</MenuItem>
          <MenuItem as={Form.Checkbox} value="green">Green</MenuItem>
          <MenuItem as={Form.Checkbox} value="yellow">Yellow</MenuItem>
        </Menu>
      </Form.CheckboxSet>
    </Form.Group>
  );
}
```

## With Modal

A ``Menu`` can be implemented to appear inside a `modalpopup` for a wide variety of use cases. The ``Modal`` brings focus to the first menu element upon the click of the trigger, and can be escaped on click away or key press.

```jsx live
() => {
  const [isOpen, open, close] = useToggle(false);
  const [target, setTarget] = React.useState(null);
  const [selected, setSelected] = useState('...');

  return (
    <>
      <Badge variant="secondary">I like {selected}</Badge>
      <Button ref={setTarget} variant="primary" size="inline" onClick={open}>Click Me To Pick:</Button>
      <ModalPopup positionRef={target} isOpen={isOpen} onClose={close} style={{
        width: 500, height: 50
      }}>
      <div className="bg-white">
        <Menu>
          <MenuItem as={Button} variant="tertiary" size="inline" onClick= {()=>setSelected('Beans')}>Beans</MenuItem>
          <MenuItem as={Button} variant="tertiary" size="inline" onClick= {()=>setSelected('Greens')}>Greens</MenuItem>
          <MenuItem as={Button} variant="tertiary" size="inline" onClick= {()=>setSelected('Tomatoes')}>Tomatoes</MenuItem>
          <MenuItem as={Button} variant="tertiary" size="inline" onClick= {()=>setSelected('Potatoes')}>Potatoes</MenuItem>
        </Menu>
      </div>
      </ModalPopup>
    </>
  )
}
```
