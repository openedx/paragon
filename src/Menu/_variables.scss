$menu-box-shadow:                                $box-shadow !default;
$menu-border-radius:                             .25em !default;
$menu-bg:                                        $white !default;
$menu-max-height:                                16.813rem !default;
$menu-item-height:                               3rem !default;
$menu-item-width:                                19rem !default;
$menu-item-width-xs:                             13.438rem !default;
$menu-item-color:                                $body-color !default;
$menu-item-bg:                                   transparent !default;
$menu-item-border-color:                         $menu-item-bg !default;
$menu-item-border-width:                         $btn-border-width !default;
$menu-item-hover-border:                         $menu-item-bg !default;
$menu-item-padding-x:                            $btn-padding-x !default;
$menu-item-padding-y:                            $btn-padding-y !default;
$menu-item-hover-color:                          $btn-tertiary-color !default;
$menu-item-hover-border-color:                   $menu-item-bg !default;
$menu-item-hover-bg:                             $btn-tertiary-hover-bg !default;
$menu-item-icon-margin-left:                     .25em !default;
$menu-item-icon-margin-right:                    $menu-item-icon-margin-left !default;
$menu-item-focus-bg:                             $btn-tertiary-active-bg !default;
$menu-item-border:                               $menu-item-border-width solid $menu-item-border-color !default;
$menu-select-btn-link-color:                     $primary-500 !default;
$menu-select-btn-link-text-decoration-line:      underline !default;
$menu-select-btn-link-text-decoration-thickness: .125rem !default;
