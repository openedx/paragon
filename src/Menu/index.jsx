import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import useArrowKeyNavigationHook from '../hooks/useArrowKeyNavigation';

const Menu = ({
  as,
  arrowKeyNavigationSelector,
  children,
  ...props
}) => {
  const parentRef = useArrowKeyNavigationHook({ selectors: arrowKeyNavigationSelector });
  const className = classNames(props.className, 'pgn__menu');

  return React.createElement(
    as,
    {
      ...props,
      ref: parentRef,
      className,
    },
    (
      <>
        <div className="bg-white">
          {children}
        </div>
      </>
    ),
  );
};

Menu.propTypes = {
  className: PropTypes.string,
  arrowKeyNavigationSelector: PropTypes.string,
  as: PropTypes.elementType,
  children: PropTypes.node,
};

Menu.defaultProps = {
  className: undefined,
  arrowKeyNavigationSelector: 'a:not(:disabled),button:not(:disabled),input:not(:disabled)',
  as: 'div',
  children: null,
};

export default Menu;
