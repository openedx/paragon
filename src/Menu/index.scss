:root {
  --pgn-elevation-menu-box-shadow:
    var(--pgn-elevation-menu-box-shadow-offset-x)
    var(--pgn-elevation-menu-box-shadow-offset-y)
    var(--pgn-elevation-menu-box-shadow-blur)
    var(--pgn-elevation-menu-box-shadow-color);
}

.pgn__menu {
  border-radius: var(--pgn-size-menu-base-border-radius);
  box-shadow: var(--pgn-elevation-menu-box-shadow);
  background-color: var(--pgn-color-menu-bg);
  overflow: auto;
  max-height: var(--pgn-size-menu-base-max-height);

  .pgn__menu-item {
    display: flex;
    align-items: center;
    justify-content: flex-start;
    width: var(--pgn-size-menu-item-width-base);
    color: var(--pgn-color-menu-item-color);
    text-align: center;
    vertical-align: middle;
    user-select: none;
    background-color: var(--pgn-color-menu-item-bg);
    border: var(--pgn-size-menu-item-border-width) solid var(--pgn-color-menu-item-border);
    padding: var(--pgn-spacing-menu-item-padding-y) var(--pgn-spacing-menu-item-padding-x);
    height: var(--pgn-size-menu-item-height);

    @media (max-width: var(--pgn-size-container-max-width-xs)) {
      width: var(--pgn-size-menu-item-width-xs);
    }

    .pgn__menu-item-text {
      display: block;
      text-overflow: ellipsis;
      white-space: nowrap;
      overflow: hidden;
    }

    .pgn__menu-item-content-spacer {
      flex-grow: 1;
    }

    @include hover {
      text-decoration: none;
      color: var(--pgn-color-menu-item-hover-color);
      border-color: var(--pgn-color-menu-item-hover-border);
      background: var(--pgn-color-menu-item-hover-bg);
    }

    // Disabled comes first so active can properly restyle
    &.disabled,
    &:disabled {
      opacity: var(--pgn-other-btn-disabled-opacity);
    }

    &.focus,
    &:focus {
      font-weight: var(--pgn-typography-font-weight-bolder);
      background-color: var(--pgn-color-menu-item-focus-bg);
    }

    .btn-icon-before {
      margin-inline-end: .5rem;
      margin-left: calc(var(--pgn-spacing-menu-item-icon-margin-left) * -1);

      [dir="rtl"] & {
        margin-right: calc(var(--pgn-spacing-menu-item-icon-margin-right) * -1);
        margin-left: .5rem;
      }
    }

    .btn-icon-after {
      margin-inline-start: .5rem;
      margin-right: calc(var(--pgn-spacing-menu-item-icon-margin-right) * -1);

      [dir="rtl"] & {
        margin-right: .5rem;
        margin-left: calc(var(--pgn-spacing-menu-item-icon-margin-left) * -1);
      }
    }

    &.pgn__form-checkbox {
      > input {
        flex-shrink: 0;  // When the menu item text is long, don't squish the checkbox, if present
      }

      > div {
        overflow: hidden;  // Ensure text gets truncated properly if needed
      }
    }
  }
}

.pgn__menu-select {
  .btn.btn-link {
    color: var(--pgn-color-menu-select-btn-link-color);

    &:hover,
    &:focus {
      text-decoration-line: var(--pgn-typography-menu-select-btn-link-text-decoration-line);
      text-decoration-thickness: var(--pgn-typography-menu-select-btn-link-text-decoration-thickness);
    }
  }

  .pgn__menu-select-popup {
    position: relative;
  }
}
