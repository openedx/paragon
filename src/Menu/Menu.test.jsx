import React from 'react';
import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import userEvent from '@testing-library/user-event';

import { Add, Check } from '../../icons';
import { MenuItem } from '..';
import Menu from '.';
import Hyperlink from '../Hyperlink';
import Button from '../Button';
import Form from '../Form';

const MENU_ITEM_TEXT = 'Cant Touch This';

describe('Menu Item renders correctly', () => {
  it('renders as just a div with empty usage', () => {
    render(<Menu data-testid="menu" />);
    expect(screen.getByTestId('menu')).toBeInTheDocument();
  });

  it('renders as expected with menu items', () => {
    const tree = renderer.create((
      <Menu>
        <MenuItem> A Menu Item</MenuItem>
        <MenuItem iconBefore={Add} stoven>A Menu Item With an Icon Before</MenuItem>
        <MenuItem iconAfter={Check}>A Menu Item With an Icon After </MenuItem>
        <MenuItem disabled>A Disabled Menu Item</MenuItem>
        <MenuItem as={Hyperlink} destination="https://en.wikipedia.org/wiki/Hyperlink">A Link Menu Item</MenuItem>
        <MenuItem as={Button} variant="primary" size="inline">A Button Menu Item</MenuItem>
        <MenuItem as={Form.Checkbox}>A Checkbox Menu Item</MenuItem>
      </Menu>
    )).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Renders disabled menu items, but you can\'t click them', async () => {
    const clickFn = jest.fn();
    render(
      <Menu>
        <MenuItem
          as={Button}
          iconBefore={Add}
          onClick={clickFn}
          disabled
        >
          {MENU_ITEM_TEXT}
        </MenuItem>
      </Menu>,
    );

    const button = screen.getByText(MENU_ITEM_TEXT);
    await userEvent.click(button);
    expect(clickFn).toHaveBeenCalledTimes(0);
  });
});

describe('Keyboard Interactions', () => {
  beforeEach(() => {
    render(
      <Menu>
        <MenuItem>Default</MenuItem>
        <MenuItem as={Button} iconBefore={Add}>{MENU_ITEM_TEXT}</MenuItem>
        <MenuItem iconBefore={Add}>Icon Before</MenuItem>
      </Menu>,
    );
  });

  it('should focus on the first item after click', async () => {
    const defaultItem = screen.getByText('Default').parentElement;
    await userEvent.click(defaultItem);
    expect(defaultItem).toHaveFocus();
  });

  it('should focus the next item after ArrowDown keyDown', () => {
    const defaultItem = screen.getByText('Default');
    const cantTouchThisItem = screen.getByText(MENU_ITEM_TEXT).parentElement;

    userEvent.type(defaultItem, '{arrowdown}');

    expect(cantTouchThisItem).toHaveFocus();
  });

  it('should focus the next item after Tab keyDown', () => {
    const defaultItem = screen.getByText('Default').parentElement;
    const cantTouchThisItem = screen.getByText(MENU_ITEM_TEXT).parentElement;
    defaultItem.focus();
    userEvent.tab();

    expect(cantTouchThisItem).toHaveFocus();
  });

  it('should loop focus to the first item after Tab keyDown on last item', () => {
    const defaultItem = screen.getByText('Default').parentElement;
    const iconBeforeItem = screen.getByText('Icon Before');
    iconBeforeItem.focus();
    userEvent.tab();

    expect(defaultItem).toHaveFocus();
  });

  it('should loop focus to the last item after ArrowUp keyDown on first item', () => {
    const defaultItem = screen.getByText('Default').parentElement;
    const iconBeforeItem = screen.getByText('Icon Before').parentElement;
    defaultItem.focus();
    userEvent.type(defaultItem, '{arrowup}');

    expect(iconBeforeItem).toHaveFocus();
  });

  it('should focus the previous item after Shift + Tab keyDown', () => {
    const button1 = screen.getAllByRole('button')[0];
    const button2 = screen.getAllByRole('button')[1];

    button2.focus();
    userEvent.tab({ shift: true });

    expect(button1).toHaveFocus();
  });
});
