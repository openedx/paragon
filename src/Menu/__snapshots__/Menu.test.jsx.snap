// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu Item renders correctly renders as expected with menu items 1`] = `
<div
  className="pgn__menu"
>
  <button
    className="pgn__menu-item"
  >
    <span
      className="pgn__menu-item-text"
    >
       A Menu Item
    </span>
    <span
      className="pgn__menu-item-content-spacer"
    />
  </button>
  <button
    className="pgn__menu-item"
    stoven={true}
  >
    <span
      className="pgn__icon btn-icon-before"
    >
      <svg
        aria-hidden={true}
        fill="none"
        focusable={false}
        height={24}
        role="img"
        viewBox="0 0 24 24"
        width={24}
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
          fill="currentColor"
        />
      </svg>
    </span>
    <span
      className="pgn__menu-item-text"
    >
      A Menu Item With an Icon Before
    </span>
    <span
      className="pgn__menu-item-content-spacer"
    />
  </button>
  <button
    className="pgn__menu-item"
  >
    <span
      className="pgn__menu-item-text"
    >
      A Menu Item With an Icon After 
    </span>
    <span
      className="pgn__menu-item-content-spacer"
    />
    <span
      className="pgn__icon btn-icon-after"
    >
      <svg
        aria-hidden={true}
        fill="none"
        focusable={false}
        height={24}
        role="img"
        viewBox="0 0 24 24"
        width={24}
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M9 16.2L4.8 12l-1.4 1.4L9 19 21 7l-1.4-1.4L9 16.2z"
          fill="currentColor"
        />
      </svg>
    </span>
  </button>
  <button
    className="pgn__menu-item"
    disabled={true}
  >
    <span
      className="pgn__menu-item-text"
    >
      A Disabled Menu Item
    </span>
    <span
      className="pgn__menu-item-content-spacer"
    />
  </button>
  <a
    className="pgn__hyperlink default-link standalone-link pgn__menu-item"
    href="https://en.wikipedia.org/wiki/Hyperlink"
    onClick={[Function]}
    target="_self"
  >
    <span
      className="pgn__menu-item-text"
    >
      A Link Menu Item
    </span>
    <span
      className="pgn__menu-item-content-spacer"
    />
  </a>
  <button
    className="pgn__menu-item btn btn-primary btn-inline"
    disabled={false}
    type="button"
  >
    <span
      className="pgn__menu-item-text"
    >
      A Button Menu Item
    </span>
    <span
      className="pgn__menu-item-content-spacer"
    />
  </button>
  <div
    className="pgn__form-checkbox pgn__menu-item"
  >
    <input
      className="pgn__form-checkbox-input"
      id="form-field1"
      type="checkbox"
    />
    <div>
      <label
        className="pgn__form-label"
        htmlFor="form-field1"
      >
        <span
          className="pgn__menu-item-text"
        >
          A Checkbox Menu Item
        </span>
        <span
          className="pgn__menu-item-content-spacer"
        />
      </label>
    </div>
  </div>
</div>
`;
